File: core/src/com/xenoage/zong/core/util/VoiceElementIterator.java
package com.xenoage.zong.core.util;

import static com.xenoage.utils.math.Fraction._0;
import static com.xenoage.zong.core.position.MP.mp;
import static com.xenoage.zong.core.position.MP.mp0;
import static com.xenoage.zong.core.position.MP.unknownMp;

import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

import lombok.Getter;

import com.xenoage.zong.core.Score;
import com.xenoage.zong.core.music.VoiceElement;
import com.xenoage.zong.core.position.MP;

/**
 * An iterator over all {@link VoiceElement}s in a score.
 * 
 * Iterates over the staves, measures, voices and its voice elements.
 * 
 * @author Andreas Wenger
 */
public class VoiceElementIterator
	implements Iterator<VoiceElement>, Iterable<VoiceElement> {
	
	private final Score score;
	
	@Getter private MP mp = Companion.getUnknownMp();
	
	private List<VoiceElement> elements;
	private MP nextMp = Companion.getMp0();

	
	public VoiceElementIterator(Score score) {
		this.score = score;
		elements = score.getVoice(nextMp).getElements();
		findNext();
	}

	@Override public boolean hasNext() {
		return nextMp != null;
	}

	@Override public VoiceElement next() {
		if (!hasNext())
			throw new NoSuchElementException();
		mp = nextMp;
		VoiceElement element = elements.get(nextMp.getElement());
		nextMp = nextMp.withElement(nextMp.getElement() + 1);
		nextMp = nextMp.withBeat(nextMp.getBeat().add(element.getDuration()));
		findNext();
		return element;
	}
	
	private void findNext() {
		while (true) {
			if (nextMp.getElement() < elements.size()) {
				//next element within voice exists
				break;
			}
			else {
				if (nextMp.getVoice() + 1 < score.getMeasure(nextMp).getVoices().size()) {
					//next measure within staff
					nextMp = Companion.mp(nextMp.getStaff(), nextMp.getMeasure(), nextMp.getVoice() + 1, Companion.get_0(), 0);
				}
				else if (nextMp.getMeasure() + 1 < score.getMeasuresCount()) {
					//next measure within staff
					nextMp = Companion.mp(nextMp.getStaff(), nextMp.getMeasure() + 1, 0, Companion.get_0(), 0);
				}
				else if (nextMp.getStaff() + 1 < score.getStavesCount()) {
					//next staff
					nextMp = Companion.mp(nextMp.getStaff() + 1, 0, 0, Companion.get_0(), 0);
				}
				else {
					//finished
					nextMp = null;
					break;
				}
				elements = score.getVoice(nextMp).getElements();
			}
		}
	}

	@Override public Iterator<VoiceElement> iterator() {
		return this;
	}
	
	@Override public void remove() {
		throw new UnsupportedOperationException();
	}

}

File: core/src/com/xenoage/zong/core/util/VoiceElementIterator.kt
package com.xenoage.zong.core.util

import com.xenoage.utils.math._0
import com.xenoage.zong.core.Score
import com.xenoage.zong.core.music.VoiceElement
import com.xenoage.zong.core.position.MP
import com.xenoage.zong.core.position.MP.Companion.mp0
import com.xenoage.zong.core.position.MP.Companion.unknownMp
import java.util.*

/**
 * An iterator over all [VoiceElement]s in a score.
 *
 * Iterates over the staves, measures, voices and its voice elements.
 */
class VoiceElementIterator(
		private val score: Score
) : Iterator<VoiceElement>, Iterable<VoiceElement> {

	var mp = unknownMp

	private var elements: List<VoiceElement>? = null
	private var nextMp: MP? = mp0


	init {
		elements = score.getVoice(nextMp!!).elements
		findNext()
	}

	override fun hasNext(): Boolean {
		return nextMp != null
	}

	override fun next(): VoiceElement {
		if (!hasNext())
			throw NoSuchElementException()
		mp = nextMp!!
		val element = elements!![mp.element]
		nextMp = mp.copy(element = nextMp!!.element + 1, beat = mp.beat!! + element.duration)
		findNext()
		return element
	}

	private fun findNext() {
		while (true) {
			if (nextMp!!.element < elements!!.size) {
				//next element within voice exists
				break
			} else {
				if (nextMp!!.voice + 1 < score.getMeasure(nextMp!!).voices.size) {
					//next measure within staff
					nextMp = MP(nextMp!!.staff, nextMp!!.measure, nextMp!!.voice + 1, _0, 0)
				} else if (nextMp!!.measure + 1 < score.measuresCount) {
					//next measure within staff
					nextMp = MP(nextMp!!.staff, nextMp!!.measure + 1, 0, _0, 0)
				} else if (nextMp!!.staff + 1 < score.stavesCount) {
					//next staff
					nextMp = MP(nextMp!!.staff + 1, 0, 0, _0, 0)
				} else {
					//finished
					nextMp = null
					break
				}
				elements = score.getVoice(nextMp!!).elements
			}
		}
	}

	override fun iterator(): Iterator<VoiceElement> {
		return this
	}

}
